### PR TITLE
TEP-0127: Larger Results via Sidecar Logs

### PR DESCRIPTION
**Pull request by jerop**
Today, `Results` have a size limit of 4KB per `Step` and 12KB per `TaskRun` in the best case - see [issue][4012].

The goal of [TEP-0086][tep-0086] is to support larger `Results` beyond the current size limits. TEP-0086 has many alternatives but no proposal. This TEP proposes experimenting with one of the alternatives - `Sidecar` logs. This allows us to support larger `Results` that are stored within `TaskRun` CRDs.

Huge shout out to @chitrangpatel for implementing the [POC](https://github.com/tektoncd/pipeline/pull/5695) - see the [demo](https://drive.google.com/file/d/1NrWudE_XBqweomiY24DP2Txnl1yN0yD9/view?usp=sharing) 👏 

[4012]: https://github.com/tektoncd/pipeline/issues/4012
[4808]: https://github.com/tektoncd/pipeline/issues/4808
[tep-0086]: https://github.com/tektoncd/community/blob/main/teps/0086-changing-the-way-result-parameters-are-stored.md

### CODE REVIEW

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+```shell
+kubectl patch cm feature-flags -n tekton-pipelines -p '{"data":{"max-result-size":"<VALUE-IN-BYTES>"}}'
+```
+
+Even though the size limit per `Result` is configurable, the size of `Results` are limited by CRD size limit of 1.5MB.
+If the size of `Results` exceeds this limit, then the `TaskRun` will fail with a message indicating the size limit has
+been exceeded.
+
+### Script in `Sidecar`
```
Comment (chitrangpatel) [inline teps/0127-larger-results-via-sidecar-logs.md:151]: I don't know if this matters but we upgraded to using a ko-built `binary` (similar to the entrypoint binary) instead of a bash script that I originally used. This allowed me to unit test it and uncover and fix edge cases.

Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:151]: That's great to hear, thanks for the update!

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+The `Sidecar` will run a script that:
+- is auto-generated based on the required `Results` - identified from `taskSpec.results` field.
+- periodically checks for files in the directory `/tekton/results`.
+- when a `Result` is found, it prints it to stdout in a parsable pattern.
+- when all the expected `Results` are found, it breaks out of the periodic loop and exits.
```
[the remark this answers is not in the document]
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md]: Updated the TEP to reflect the current implementation, thank you @chitrangpatel!

**Comment by skaegi** [comment]
I've read the TEP this looks very good to me! This is an ongoing problem that our users encounter. We're happy to sign up to test and use this approach. Thank-you @jerop and @chitrangpatel

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+This TEP builds on the hard work of many people who have been tackling the problem over the past couple years,
+including but not limited to:
+- '@tlawrie'
+- '@imjasonh'
+- '@bobcatfish'
+- '@pritidesai'
+- '@tomcli'
+- '@ScrapCodes'
+- '@wlynch'
```
Comment (jerop) [inline teps/0127-larger-results-via-sidecar-logs.md:43]: Lifted these names from the authors of [TEP-0086](https://github.com/tektoncd/community/blob/main/teps/0086-changing-the-way-result-parameters-are-stored.md), please let me know if you'd like to be added or removed

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+## Summary
+
+Today, `Results` have a size limit of 4KB per `Step` and 12KB per `TaskRun` in the best case - see [issue][4012]. 
+
+The goal of [TEP-0086][tep-0086] is to support larger `Results` beyond the current size limits. TEP-0086 has many
+alternatives but no proposal. This TEP is proposes experimenting with one of the alternatives - `Sidecar` logs.
+This allows us to support larger `Results` that are stored within `TaskRun` CRDs.
```
Comment (abayer) [inline teps/0127-larger-results-via-sidecar-logs.md:51]: This sentence confuses me a bit - was that supposed to be `than are stored` rather than `that are stored`?

Comment (abayer) [reply teps/0127-larger-results-via-sidecar-logs.md:51]: Oh, right, it does mean `that are stored`. I got momentarily confused. I'd phrase that as `which are stored`, fwiw. =)

Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md]: Changed `that` to `which`...English 😅 

Thank you!

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+The following are out of scope for this TEP:
+
+1. Solving use cases that requires really large `Results` beyond the size limit of a CRD (1.5MB).
+
+2. Addressing other [alternatives][tep-0086-alts] for larger `Results` that are listed in [TEP-0086][tep-0086].
+
+### Use Cases
+
+1. Store larger `Results`, such as JSON payloads from an HTTP call or a Cloud Event, that can be inspected later or
```
Comment (bobcatfish) [inline teps/0127-larger-results-via-sidecar-logs.md:90]: is it reasonable to assume these kinds of payloads will always be less than 1.5 MB? if not i wonder if this use case makes sense for what we're describing here

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+   passed to other `Tasks` without the need to understand storage mechanisms.
+
+3. The [documented guidance][docs] is that `Results` are used for outputs less than 1KB while `Workspaces` are used
```
Comment (bobcatfish) [inline teps/0127-larger-results-via-sidecar-logs.md:93]: i'm a bit confused by this use case - this seems more like justification or maybe a requirement?

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+   for larger data.
+
+   > As a general rule-of-thumb, if a `Result` needs to be larger than a kilobyte, you should likely use a `Workspace`
+     to store and pass it between `Tasks` within a `Pipeline`.
+   
+   Supporting larger `Results` upto the CRD limit allows users to reuse `Tasks` in more scenarios without having to
+   change the specification to use `Workspaces` upon hitting the current low size limit of 4KB per `TaskRun`.
+
```
Comment (bobcatfish) [inline teps/0127-larger-results-via-sidecar-logs.md:101]: the structured results use case from https://github.com/tektoncd/community/blob/main/teps/0086-changing-the-way-result-parameters-are-stored.md#use-cases-optional might make sense here

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+In the experiment, we will continue to measure the startup overhead and explore ways to improve it.
+
+For further details, see the [performance metrics][performance].
+
+### Security
+
+This approach requires that the `Pipeline` controller has access to `Pod` logs. The `Pipeline` controller already has
+extensive permissions in the cluster, such as read access to `Secrets`. Expanding the access even further is a concern
+for some users, but is also acceptable for some users given the advantages. We will document the extended permissions
+so that users can make the right choice for their own use cases and requirements.
```
Comment (bobcatfish) [inline teps/0127-larger-results-via-sidecar-logs.md:215]: not sure if this is important or not: we might want to make sure that users cant inject their own sidecar overtop of the tekton specified one

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+### Reusability
+
+This proposal does not introduce any API changes to specification `Results`. The changes are in implementation details
+of `Results`. The existing `Tasks` will continue to function as they are, only that they can support larger `Results`.
+
+Even more, supporting larger `Results` upto the CRD limit allows users to reuse `Tasks` in more scenarios without
+having to change the specification to use `Workspaces` upon hitting the current low size limit of 4KB per `TaskRun`.
+This allows users to control execution as needed by their context without having to modify `Tasks` and `Pipelines`.
+
```
Comment (bobcatfish) [inline teps/0127-larger-results-via-sidecar-logs.md:192]: One risk that might make sense to mention here is around Tasks being written that assume larger result support which would only work on Pipelines installations configured to support it - i.e. this could be a risk to Task interoperability which hopefully is mitigated by the facts that a) there is still a hard limit (CRD size) and b) we plan in the long run to support larger results regardless

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+For `ChildReferences` to be populated, the `embedded-status` must be set to `"minimal"`. We recommend that the minimal
+embedded status - `ChildReferences` - is enabled while migration is ongoing until it becomes the only supported status.
+This will ensure that larger `Results` from all its `TaskRuns` will not bloat the `PipelineRun` CRD.
+
+## Design Details
+
+The `Sidecar` will run a binary that:
+- receives argument for `Results`' paths and names which are identified from `taskSpec.results` field. This allows the
+  `Sidecar` to know the `Results` it needs to read.
+- has `/tekton/run` volume mounted where status of each `Step` is written.
```
Comment (bobcatfish) [inline teps/0127-larger-results-via-sidecar-logs.md:175]: nit: we probably want to make sure the volumes are mounted as read only (https://github.com/tektoncd/pipeline/issues/4227)

**Review by bobcatfish — COMMENTED** [review]
Some minor comments from me but generally I'm very much in favor and excited to see this move forward! 🎉 

/approve

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 10 above)
Comment (chitrangpatel) [reply teps/0127-larger-results-via-sidecar-logs.md:175]: yes they are :). We should probably mention that.

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 8 above)
Comment (chitrangpatel) [reply teps/0127-larger-results-via-sidecar-logs.md:215]: great catch! I can add a validation check so that the webhook throws an error if it finds a matching sidecar name as that reserved by the reconciler in case they set their result extraction method as sidecar-logs.

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 5 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:90]: yes, we can't make that assumption so replaced this use case with:

- the non-falsifiable provenance use case where the signatures and certificates are stored alongside the results to be verified -- https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md#signing-results
- publishing container images alongside all their copies in regional registries as we experienced in the dogfooding cluster - https://github.com/tektoncd/pipeline/issues/4282

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 6 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:93]: yes, more of a justification, so i moved it to the motivation section

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 7 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:101]: ah makes sense, just seeing the comment suggesting the "structured results use case" -- updated the use case with details about being "structured"

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 9 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:192]: yes yes, added a new "interoperability" section with this concern and how we mitigate it

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 10 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:175]: mentioned that it's read only 👍🏾

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 8 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:215]: added a note about this, thank you @bobcatfish and @chitrangpatel!

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+The main goal of this TEP is to support larger `Results` via `Sidecar` logs. The `Results` are stored in the `TaskRun`,
+therefore they are limited by the size limit of a `TaskRun` CRD - 1.5MB. 
+
+### Non-Goals
+
+The following are out of scope for this TEP:
+
+1. Solving use cases that requires really large `Results` beyond the size limit of a CRD - 1.5MB.
+
+2. Addressing other [alternatives][tep-0086-alts] for larger `Results` that are listed in [TEP-0086][tep-0086].
```
Comment (afrittoli) [inline teps/0127-larger-results-via-sidecar-logs.md:96]: This is fine of course, but I wonder if we should include in the goals proposing a solution that may cohexists with other solutions from the alternatives in TEP-0086, for instance via feature flag to being with.

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+functionality the same and does not require any external storage mechanism.
+
+For further details, see the [demonstration][demo] of the [implementation][poc].
+
+This proposal provides an opportunity to experiment with this solution to provide `Results` within the CRDs as we
+continue to explore other alternatives, including those that involve external storage.
+
+### Feature Gate
+
+This feature will be gated using a `result-extraction-method` feature flag.
```
Comment (afrittoli) [inline teps/0127-larger-results-via-sidecar-logs.md:137]: NIT: may I suggest something simpler, like `result-storage` or `results-from`?

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+```shell
+kubectl apply -f config/enable-log-access-to-controller/
+```
+
+### Size Limit
+
+The size limit per `Result` can be configured using the `max-result-size` feature flag, which takes the integer value
+of the bytes.
+
+The `max-result-size` feature flag defaults to 4096 bytes. This ensures that we support existing `Tasks` with only one
```
Comment (afrittoli) [inline teps/0127-larger-results-via-sidecar-logs.md:160]: +1 - thank you!

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+`Result` that uses up the whole size limit of the `Termination Message`.
+
+If users want to set the size limit per `Result` to be something other than 4096 bytes, they can set `max-result-size`
+by setting `max-result-size: <VALUE-IN-BYTES>`. The value set here cannot exceed the CRD size limit of 1.5MB.
```
Comment (afrittoli) [inline teps/0127-larger-results-via-sidecar-logs.md:164]: What happens in case the value is wrong? 
Does the controller log an error and adopt the default value?

**Comment by afrittoli** [comment]
A couple of comments, but nothing blocking.
Thank you for this, looking forward to this feature!
/approve

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 11 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:96]: Agreed that they should coexist, updated the TEP

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 14 above)
Comment (chitrangpatel) [reply teps/0127-larger-results-via-sidecar-logs.md:164]: yes. That's right.

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 12 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:137]: `results-from` is much better name...updated, thank you!

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 14 above)
Comment (chitrangpatel) [reply teps/0127-larger-results-via-sidecar-logs.md:164]: there will be an error in the controller when the config is applied and the default value is used.

File: teps/0127-larger-results-via-sidecar-logs.md [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+### Feature Gate
+
+This feature will be gated using a `result-extraction-method` feature flag. 
+
+Users have to set `result-extraction-method` to `"sidecar-logs"` to enable the feature:
+```shell
+kubectl patch cm feature-flags -n tekton-pipelines -p '{"data":{"result-extraction-method":"sidecar-logs"}}'
+```
+
+This feature is disabled by default. When disabled, `Results` will continue to pass through `Termination Message`.
```
[the remark this answers is not in the document]
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:144]: great idea, updated the TEP to reflect that 👍🏾

File: teps/0127-larger-results-via-sidecar-logs.md (hunk 14 above)
Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md:164]: I'd rather we log an error and fail, instead of adopting the default value, so that it the behavior is not surprising when a user thinks they set a higher limit but then the task fails later capped by the default limit

Updated the TEP to make it clear that we'll validate the value and fail if it's incorrect

Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md]: @afrittoli @chitrangpatel are you okay with not using the default, and failing instead?

Comment (chitrangpatel) [reply teps/0127-larger-results-via-sidecar-logs.md]: hmm. I'm not sure how to fail. This is a value in config-feature-flags, not task/pipeline spec. So it's applied to the controller regardless of which task/pipeline is run.

Comment (chitrangpatel) [reply teps/0127-larger-results-via-sidecar-logs.md]: May be I'm misunderstanding. I'm happy to meet and discuss it as well.

Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md]: was thinking of how we validate the values in defaults.go, for example: https://github.com/tektoncd/pipeline/blob/8c59d00342e9479e68e9c52f6d18b62ac3015489/pkg/apis/config/default.go#L147-L153

Comment (jerop) [reply teps/0127-larger-results-via-sidecar-logs.md]: or we validate this also in feature_flags.go, for example:

https://github.com/tektoncd/pipeline/blob/c2d028378db43ac44d19c10bb68102dd057fbe2e/pkg/apis/config/feature_flags.go#L210-L224
